src/gradient.jl

```diff
@@ -28,6 +28,11 @@ end
 
 @inline gradient!(out, f::F, x::SArray) where {F} = vector_mode_gradient!(out, f, x)
 
+@inline function gradient!(out::ImmutableDiffResult, f, x::SArray)
+    dual = vector_mode_dual_eval(f, x)
+    out = DiffResult(value(dual), extract_gradient(dual, x))
+end
```

Comment:
I think the same is true here (and similar methods) - these methods shouldn't be necessary, since the `DiffBase` API should do all this work for us as long as we switch over to the new output-aliasing convention.